### PR TITLE
Security fix for XML parser

### PR DESCRIPTION
**Pull request by samathaamz**
### Summary

XML Parser security fix

### Description

- Added new configuration parameter `allowXmlUnsafeExternalEntity` 
- Edited README to document the new configuration parameter
- Added corresponding unit tests

### Additional Reviewers

@seneramz 
@samathaamz 
@sergiyvamz 
@hsuamz 
@congoamz

### CODE REVIEW

File: src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java [hunk 1]
```diff
@@ -662,6 +662,9 @@
                 new BooleanPropertyDefinition(PropertyKey.acceptAwsProtocolOnly, DEFAULT_VALUE_FALSE, RUNTIME_NOT_MODIFIABLE,
                         Messages.getString("ConnectionProperties.acceptAwsProtocolOnly"), "0.1.0", CATEGORY_HA, Integer.MAX_VALUE),
 
+                new BooleanPropertyDefinition(PropertyKey.allowXmlUnsafeExternalEntity, DEFAULT_VALUE_FALSE, RUNTIME_NOT_MODIFIABLE,
+                        Messages.getString("ConnectionProperties.allowXmlUnsafeExternalEntity"), "0.2.0", CATEGORY_HA, Integer.MAX_VALUE),
```
Comment (sergiyvamz) [inline src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java:666]: I don't think it's CATEGORY_HA. It should be something generic.

File: src/main/user-impl/java/com/mysql/cj/jdbc/MysqlSQLXML.java [hunk 2]
```diff
@@ -212,6 +218,16 @@ public synchronized Reader getCharacterStream() throws SQLException {
             try {
                 DocumentBuilderFactory builderFactory = DocumentBuilderFactory.newInstance();
                 builderFactory.setNamespaceAware(true);
+
+                if (!this.allowXmlUnsafeExternalEntity) {
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/MysqlSQLXML.java:222]: Can we have few unit tests to make sure it fixes the issue?

Comment (samathaamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/MysqlSQLXML.java:222]: Yeap for sure!

File: src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java (hunk 1 above)
Comment (samathaamz) [reply src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java:666]: Would CATEGORY_JDBC be better or CATEGORY_SECURITY or something else?

Comment (sergiyvamz) [reply src/main/core-api/java/com/mysql/cj/conf/PropertyDefinitions.java:666]: CATEGORY_SECURITY seems a good option.

File: src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+import javax.xml.transform.dom.DOMSource;
+
+public class SQLXMLTests {
+
+    @Test
+    public void testAllowXmlUnsafeExternalEntityPropertyFalse() {
+
+        final String url =
+          "jdbc:mysql:aws://somehost:1234/test";
```
Comment (sergiyvamz) [inline src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java:23]: "jdbc:mysql://"?

File: src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        final ConnectionUrl conStr = ConnectionUrl.getConnectionUrlInstance(url, new Properties());
+        JdbcPropertySetImpl connProps1 = new JdbcPropertySetImpl();
```
Comment (sergiyvamz) [inline src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java:25]: connProperties?

File: src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        connProps1.initializeProperties(conStr.getMainHost().exposeAsProperties());
+
+        MysqlSQLXML xmlTest = new MysqlSQLXML(null, connProps1);
+
+        assertTrue(xmlTest.getAllowXmlUnsafeExternalEntity());
+
+    }
+
+    @Test
+    public void testXXEInjectionPrevention() throws SQLException {
```
Comment (sergiyvamz) [inline src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java:74]: Can we have similar test with allowXmlUnsafeExternalEntity=true?

File: src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+        MysqlSQLXML xmlTest = new MysqlSQLXML(null, connProperties);
+
+        xmlTest.setString(xmlString);
+        SQLException exception = assertThrows( SQLException.class, () -> xmlTest.getSource(DOMSource.class));
+
+        // Assert correct error code was produced
+        assertSame("S1009", exception.getSQLState());
+        // Assert that the reason for this exception is because the file does not exist (and not because DTDs is disabled)
+        assertTrue(expectedErrorMsg.equals(exception.getMessage()));
```
Comment (sergiyvamz) [inline src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java:123]: It fails in Windows. Exception message is
\dev\nonExistantFile (The system cannot find the path specified)

It's probably better to check for an expected substring in exception message.

Comment (samathaamz) [reply src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java]: I just pushed a changed but just realized even the text is completely different.

Comment (samathaamz) [reply src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java]: Changed it to check if the string contains either "(The system cannot find the path specified)" or "(No such file or directory)"

Comment (samathaamz) [reply src/test/java/com/mysql/cj/jdbc/SQLXMLTests.java]: Another option we could do is to check if the error message is different than the previous test.

**Pull request merged**